src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java

```diff
@@ -0,0 +1,392 @@
+          ZimbraLog.extensions.debug("Application limit reached: " + errorMsg);
+          throw ServiceException.TEMPORARILY_UNAVAILABLE();
+        case FacebookConstants.RESPONSE_ERROR_POLICIES_VIOLATION:
+          ZimbraLog.extensions.debug("Temporarily blocked for policies violations: " + errorMsg);
+          throw ServiceException.TEMPORARILY_UNAVAILABLE();
+        case FacebookConstants.RESPONSE_ERROR_TOO_MANY_REQUESTS:
+          ZimbraLog.extensions.debug("Too many requests: " + errorMsg);
+          throw ServiceException.TEMPORARILY_UNAVAILABLE();
+        default:
+          ZimbraLog.extensions.warn("Unexpected error while trying to authenticate "
+              + "the user." + response.toString());
+          throw ServiceException.PERM_DENIED("Unable to authenticate the user.");
+      }
+    }
+  }
+
+  /**
+   * Check code and return range which is listed in the FacebookConstants class.<br>
+   * See Facebook error code range for Permission denied errors<br>
+   * https://developers.facebook.com/docs/graph-api/using-graph-api/error-handling
+   *
+   * @param errorCode An error code from value
+   * @return The Range that matches the error type
+   */
+  protected String inErrorCodeRange(String errorCode) {
+    if (!errorCode.isEmpty()) {
+      Integer errorCodeInt = Integer.getInteger(errorCode);
+      if (errorCodeInt >= 200 && errorCodeInt <= 299) {
+        errorCode = "200-299";
+      }
+    }
+    return errorCode;
+  }
+
+  /**
+   * Retrieves the primary email of the user with the access_token and auth
+   * token.
+   *
+   * @param credentials The json containing an access_token
+   * @return The unique user ID
+   * @throws ServiceException If there are issues determining the primary address
+   */
+  @Override
+  protected String getPrimaryEmail(JsonNode credentials) throws ServiceException {
+    JsonNode json = null;
+    final String authToken = credentials.get("access_token").asText();
+    String url = FacebookConstants.DEBUG_TOKEN_URI;
+    String queryString;
+    try {
+      queryString = "?input_token=" + authToken + "&access_token="
+          + URLEncoder.encode(getAppToken(), OAuth2Constants.ENCODING);
+    } catch (UnsupportedEncodingException e1) {
+      throw ServiceException.PARSE_ERROR("Url encoding the Facebook app token failed.", e1);
+    }
+
+    try {
+      final GetMethod request = new GetMethod(url);
+      request.setQueryString(queryString);
+      json = executeRequestForJson(request);
+    } catch (final IOException e) {
+      ZimbraLog.extensions.debug("There was an issue acquiring the authorization token.", e);
+      throw ServiceException.PROXY_ERROR("There was an issue acquiring an authorization token "
+          + "for this user.", null);
+    }
+    String userId = json.get("data").get("user_id").asText();
+    if (userId.isEmpty()) {
+      throw ServiceException.NOT_FOUND("User Id not found in json data. " + json.toString());
```

Comment:
Don't put the json at the end of the error here, we don't want that in the browser. Use the same error message the other implementations use.

Reply:
If the `user_id` is not in this response, what does that imply?

In Yahoo's case it means the yahoo app is mis-configured, is it similar here or is it not even possible?